### ISSUE TITLE
Deeper investigation into DotProduct loss

### ISSUE THREAD

**Issue by dakshvar22** [RasaHQ/rasa#7698]
The `DotProduct` loss used by all the ML components can be summarized to be doing the following currently - 

1. Obtain embedding for batch of input(message / tracker state) and labels(intent / response / action): `i` and `l`.
2. Randomly sample a set of negative embeddings for both input and labels - `neg_i`, and `neg_l`
3. Compute five terms - `sim_i_pos_l`, `sim_i_neg_l`, `sim_l_neg_i`, `sim_i_neg_i`, `sim_l_neg_l`. 
Here `sim` denotes a similarity function(cosine or inner product). `sim_x_y` denotes similarity between embeddings `x` and `y`. `pos_l` is the ground truth label for the corresponding `i`.
4. Maximize `sim_i_pos_l` and minimize all other terms. How is this done inside [`_softmax_loss`](https://github.com/RasaHQ/rasa/blob/master/rasa/utils/tensorflow/layers.py#L845)?
     1. Treat all the similarity terms as logits.
     2. Apply softmax on these logits.
     3. Use cross entropy as the loss function with labels `1` and `0` for `sim_i_pos_l` and all others respectively.

What we found from experiments(these were primed by a few hypothesis) - 

1. Softmax over logits works great to maximize the separation between [`sim_i_pos_l`] and [`sim_i_neg_l`, `sim_l_neg_i`](Referred to as positive and negative pairs ahead). However, it does not constrain the original similarity values in an approximate range. That's intuitive to imagine because as long as there is a good separation between positive and negative pairs of similarity, cross entropy over softmax probabilities will be low. As an effect of this, the mean value of positive and negative pairs after the model is trained is around `+12` and `+6` respectively(validated on a couple of datasets across different configurations). These aren't really ideal values for similarities as it can have adverse effects on some proxy tasks and real world test sets.
2. Another related observation was that `sim_i_neg_i` and `sim_l_neg_l` don't need to be minimized relative to other similarity terms. In fact, doing so actually pushed the similarity values of `sim_pos` to a higher value since the model struggles to minimize `sim_i_neg_i` to lower values and hence keeps trying to push `sim_pos` higher to maximize the separation.
3. In order to constrain all similarity values to an approximate range, we applied `sigmoid` on each of them and added the cross entropy loss over sigmoid probabilities to the final total loss. This brought back mean similarity values to a better range of `+6` and `-0.5` for positive and negative pairs respectively. Along with that, logits for softmax were constructed only from `sim_i_pos_l`, `sim_i_neg_l` and `sim_l_neg_i`. These are the similarity terms which need to be maximized and minimized relative to each other. Logits for sigmoid were constructed from all 5 similarity terms and hence each of them individually constrained to a range. Quantiative results are [better](https://github.com/RasaHQ/rasa/pull/7616#issuecomment-750193667) for some datasets and [at par with master on others](https://github.com/RasaHQ/rasa/pull/7616#issuecomment-755700972)
4. Next set of experiments tried to validate the negative sampling procedure. Instead of randomly sampling negatives, we mine hard negatives by choosing those embeddings which have the highest similarity in the pair. For example, to compute `sim_i_neg_l` we choose those set of `neg_l` for which `sim_i_neg_l` turns out to be maximum. We call this strategy as the sorted sampling strategy. We evaluate this on two levels - 
    1. F1 score on test set: On large datasets like Carbon Bot and Sara, the performance is the same as random negative sampling. However, on smaller datasets with low number of examples per intent a lot of configurations benefit. On the [HINT3 dataset](https://arxiv.org/abs/2009.13833) we achieve almost the [SoTA](https://arxiv.org/pdf/2012.03929.pdf) or in some cases slightly better as well. Results can be found in [1](https://github.com/RasaHQ/rasa/pull/7536#issuecomment-748529486) and [2](https://github.com/RasaHQ/rasa/pull/7536#issuecomment-748618378). The boosts here are moderate compared to even random sampling + sigmoid and considerably large compared to current performance on master.
    2. Range of similarity values: To track this, along with the mean similarity value of positive and negative pairs we also calculate the standard deviation. Ideally we want the standard deviation to not be too high. With random negative sampling + sigmoid added to the loss term, the similarity values for the negative pairs are around `-0.3(+/-2.37)` . When employing sorted sampling strategy + sigmoid added to the loss term, the similarity values for negative pairs are around `0.72(+/-0.42)`. While the mean value is higher, standard deviation is much lower for hard negative strategy which is a good result. These numbers were on `Private 3` dataset with `sparse + DIET` config. However - 
        1. It's harder to train with sorted sampling strategy + increasing batch size as the optimization gets stuck in an oscillating behaviour on some datasets(specially small ones). For this reducing the batch size and keeping it fixed to 32 works the best.
        2. The training time is increased because of the sorting that is done on every pass of a minibatch.
5. We also found that minimizing `sim_i_neg_i` is tough. A side effect of trying to minimize it is that it impacts the model's ability to minimize similarities of other negative pairs after a certain extent. (The mean il, li similarity that is achieved after training is ~ `-0.5`). Removing `sim_i_neg_i` lets the model achieve a mean value of ~ `6.5` for `sim_i_pos_l` and ~ `-7` for `sim_i_neg_l`  and `sim_l_neg_i` which is the ideal range we would want them to be in. However, on the other hand not minimizing `sim_neg_ii` also impacts test performance. This seems like a classical problem of multi-task loss where the value of one term dominates the overall loss value and goes in a cyclic oscillating behaviour where after a point training loss and training accuracy is saturated and cannot be improved further. Two approaches that we are currently trying to see if they help - 
    1. Weigh down the contribution of `sim_i_neg_i` by multiplying it with a small constant. Tried with a constant value of `0.02` and it does seem to [benefit](https://github.com/RasaHQ/rasa/pull/7536#issuecomment-754680874)(on sorted sampling)
    2. Separate the loss into main loss (containing main softmax and maybe sigmoid) and auxiliary loss containing sim_i_neg_i sigmoid and others, then we filter the auxiliary gradients if they adhere to a certain condition. The filtering condition needs to be thought through.
    3. Use a smaller number negative samples for `sim_i_neg_i`.

TODOs from above experiments-

- [x] [PR](https://github.com/RasaHQ/rasa/pull/7616) for introducing sigmoid term inside the softmax loss.
- [ ] PR for sorted sampling strategy but turned off by default and made configurable. Advise users to lower the batch size when using it.
- [x] Look into `sim_neg_ii` and ways to mitigate the problem.

**Comment by koernerfelicia** [comment]
@dakshvar22 is this in progress? Are the TODOs dependent on each other? Are they both TODOs that can be picked up by other people or are these notes for you?

**Comment by dakshvar22** [comment]
@koernerfelicia This is in progress and both me and @Ghostvv are working together on 1st and 3rd TODO. The 2nd TODO will follow once the 1st and 3rd are finished.

**Comment by dakshvar22** [comment]
1st and 3rd TODOs of this were shipped with https://github.com/RasaHQ/rasa/pull/7616

**Comment by dakshvar22** [comment]
We don't plan to attempt the second TODO as of now because of competing priorities. Deprioritizing for now and closing the issue.
